### PR TITLE
STYLE: turn off PLW2901 (but keep some changes)

### PR DESCRIPTION
**Pull request by ErdiTk**
Continuation of work started on closed issue #51708.
Didn't find a new issue to link this PR, may I ask you @MarcoGorelli if it is fine to link it to the closed one or shall I wait for a new issue to be opened?

### CODE REVIEW

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks

can you update the pyproject.toml file accordingly please?

**Comment by ErdiTk** [comment]
> thanks
> 
> can you update the pyproject.toml file accordingly please?

I will update it when I finish all the files under  `pandas/core/*`. Give me few days for it, there are quite some to change :)

**Comment by MarcoGorelli** [comment]
it's fine (and preferable) split it up into smaller chunks

we can split up the ` pandas/core/*` ignore into multiple ones

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
looks like the "
Code Checks / Docstring validation, typing, and other manual pre-commit hooks (pull_request) " is failing - code you give me a ping when that's one's fixed please?

**Comment by ErdiTk** [comment]
> looks like the "
> Code Checks / Docstring validation, typing, and other manual pre-commit hooks (pull_request) " is failing - code you give me a ping when that's one's fixed please?

Sure, I think I know which changes might have impacted it. I will ping you around Tuesday, unfortunately I won't make it earlier.

**Comment by MarcoGorelli** [comment]
sure, no hurry, thanks!

File: pandas/core/apply.py [hunk 1]
```diff
@@ -1427,14 +1427,17 @@ def _managle_lambda_list(aggfuncs: Sequence[Any]) -> Sequence[Any]:
...
-            aggfunc = partial(aggfunc)
-            aggfunc.__name__ = f"<lambda_{i}>"
-            i += 1
-        mangled_aggfuncs.append(aggfunc)
+    for idx, aggfunc in enumerate(aggfuncs):
+        if not com.get_callable_name(aggfunc) == "<lambda>":
+            mangled_aggfuncs.append(aggfunc)
+            continue
+
+        partial_aggfunc = aggfunc
```
Comment (ErdiTk) [inline pandas/core/apply.py:1437]: @MarcoGorelli by adding this line the error should be solved. At least locally the mypy error is gone. Not sure why but by simply creating `partial_aggfunc` from the call to `partial(aggfunc)` in the next line it removes the `__name__` attribute. But if I firstly assign it as equal to `aggfunc` as I am doing here it works fine. Do you know any similar behavior?

File: pandas/core/apply.py [hunk 2]
```diff
@@ -1427,14 +1427,17 @@ def _managle_lambda_list(aggfuncs: Sequence[Any]) -> Sequence[Any]:
...
+        partial_aggfunc = partial(aggfunc)
+        partial_aggfunc.__name__ = f"<lambda_{idx}>"
```
Comment (MarcoGorelli) [inline pandas/core/apply.py:1439]: also, I don't think the logic is right here, please try to keep the logic the same as the original

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for your PR

I've left a couple of comments, there should be enough to work on here

File: pandas/core/apply.py (hunk 1 above)
Comment (ErdiTk) [reply pandas/core/apply.py]: Oh thank you a lot for the suggestion! I will implement it like this. Sorry but I am still learning how to properly handle those issues.

File: pandas/core/apply.py (hunk 2 above)
Comment (ErdiTk) [reply pandas/core/apply.py]: The original one was:
```
i = 0
for aggfunc in aggfuncs:
        if com.get_callable_name(aggfunc) == "<lambda>":
            aggfunc = partial(aggfunc)
            aggfunc.__name__ = f"<lambda_{i}>"
            i += 1
        mangled_aggfuncs.append(aggfunc)
```

I removed the variable `i`, replacing it by `idx` generated by the call to `enumerate`. Then I inverted the `if` clause, making it a `if not`. In this way I can insert inside the condition the single line:
`mangled_aggfuncs.append(aggfunc)`
and then continue the loop. So I moved what was inside the if outside. Sorry I understand it's not super clear, but I thought would look nicer to have a lighter content inside the if statement. If you want I can revert the logic.

File: pandas/core/arrays/period.py [hunk 3]
```diff
@@ -1105,7 +1105,7 @@ def _range_from_fields(
         freqstr = freq.freqstr
         year, quarter = _make_field_arrays(year, quarter)
         for y, q in zip(year, quarter):
-            y, m = parsing.quarter_to_myear(y, q, freqstr)
+            _, m = parsing.quarter_to_myear(y, q, freqstr)
             val = libperiod.period_ordinal(y, m, 1, 1, 1, 1, 0, 0, base)
```
[the remark this answers is not in the document]
Comment (ErdiTk) [reply pandas/core/arrays/period.py]: My bad, indeed the variable `y` can be changed inside the function `quarter_to_myear`. Updated! Thanks for the heads up!

File: pandas/core/apply.py [hunk 4]
```diff
@@ -1427,14 +1427,17 @@ def _managle_lambda_list(aggfuncs: Sequence[Any]) -> Sequence[Any]:
...
-        mangled_aggfuncs.append(aggfunc)
+    for idx, aggfunc in enumerate(aggfuncs):
+        if not com.get_callable_name(aggfunc) == "<lambda>":
+            mangled_aggfuncs.append(aggfunc)
+            continue
+
+        partial_aggfunc = partial(aggfunc)
+        # "partial[Any]" has no attribute "__name__"
+        partial_aggfunc.__name__ = f"<lambda_{idx}>"  # type: ignore[attr-defined]
+        mangled_aggfuncs.append(partial_aggfunc)
```
Comment (MarcoGorelli) [inline pandas/core/apply.py:1440]: this isn't the same logic - before, `i` was only incremented if `com.get_callable_name(aggfunc) == "<lambda>"`, whereas now you're incrementing `idx` on every iteration

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/period.py [hunk 5]
```diff
@@ -1105,8 +1105,8 @@ def _range_from_fields(
         freqstr = freq.freqstr
         year, quarter = _make_field_arrays(year, quarter)
         for y, q in zip(year, quarter):
-            y, m = parsing.quarter_to_myear(y, q, freqstr)
-            val = libperiod.period_ordinal(y, m, 1, 1, 1, 1, 0, 0, base)
+            y_parsed, m = parsing.quarter_to_myear(y, q, freqstr)
```
Comment (MarcoGorelli) [inline pandas/core/arrays/period.py:1108]: maybe `calendar_year, calendar_month = parsing.quarter_to_myear(y, q, freqstr)`?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for working on this

I'm going to go back to my previous comment https://github.com/pandas-dev/pandas/pull/51797#pullrequestreview-1327035638 , I'm not sure we want to enable this in CI

some of these changes are nice

shall we keep the
```
where[idx] = _validate_where(w)
```
change (which is quite nice), revert the rest, and move `PLW2901` to the 'intentionally turned off' section of pyproject.toml?

https://github.com/pandas-dev/pandas/blob/af55880968d43c907dcf9b679f1ce227d76c143e/pyproject.toml#L316

**Comment by ErdiTk** [comment]
Ok, I reverted most of the changes except on few files.
Under `# intentionally turned off ` I added `"redefined-loop-name"`, but I saw the errors still appearing when running `pre-commit run ruff --all-files`, so I kept `"PLW2901"` also under the to ignore section of ruff. Please let me know if this is the correct way. Also let me know if the few changes listed are fine.

File: pandas/core/apply.py [hunk 6]
```diff
@@ -1423,10 +1423,10 @@ def relabel_result(
         # mean  1.5
         # mean  1.5
         if reorder_mask:
-            fun = [
```
Comment (MarcoGorelli) [inline pandas/core/apply.py:1426]: not sure we can do this, `fun` is still used below

**Review by MarcoGorelli — APPROVED** [review]
looks fine, just made some minor fixup, thanks @ErdiTk

**Pull request merged**